### PR TITLE
Linux Installer Script

### PR DESCRIPTION
**Pull request by justinh-rahb**
### Description:

This Pull Request introduces a new script designed to facilitate the automatic installation of Open WebUI on Ubuntu systems. The script offers a streamlined process for setting up Open WebUI, with support for both Docker-based and direct installations. Key features of this script include:

- **Docker Installation:** For users who prefer or require a Docker-based setup, the script checks for the presence of Docker on the system. If Docker is not found, it prompts the user with the option to install Docker directly. This ensures that users can easily prepare their environment for running Open WebUI within a Docker container.

- **Open WebUI Setup:** The script handles the deployment of Open WebUI, providing options for either a Docker container setup or a direct installation. This flexibility allows users to choose the installation method that best fits their needs and system capabilities.

- **User Interaction:** Throughout the process, the script engages with the user through prompts, ensuring that they are in control of the installation choices, such as opting in for Docker installation or selecting the preferred method for deploying Open WebUI.

> [!NOTE]  
> The current version of the script is tailored for Ubuntu environments. However, work is underway to extend support to additional Linux distributions. This expansion will cater to a broader user base, accommodating different system preferences and requirements.

### Future Plans:

The development of this script is ongoing, with plans to polish existing features and introduce support for more Linux distributions. Key goals include:

- **Enhanced Distro Support:** Expanding the script to support other popular Linux distributions, making the installation process as seamless on those systems as it is on Ubuntu.

- **Streamlined User Experience:** Continuous improvements to the script's user interaction model, ensuring it remains intuitive and user-friendly.

- **Official Hosting:** In the future, we aim to host this script on `https://openwebui.com`, allowing users to easily initiate the installation process with a simple curl command:
  ```bash
  curl https://openwebui.com/install.sh | bash
  ```
  This will significantly simplify the setup process for Open WebUI, making it accessible to users with minimal setup commands.

### Request for Comments:

I invite feedback, suggestions, and contributions from the community to refine this script further. Collaboration will be key to extending its capabilities and ensuring it meets the needs of our diverse user base. If you have ideas for improvement, additional features, or support for other distributions, please feel free to contribute to this effort.

### CODE REVIEW

File: install-linux.sh [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+      "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] https://download.docker.com/linux/ubuntu \
+      $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
+    sudo apt-get update
+    sudo apt-get install -y docker-ce docker-ce-cli containerd.io
+    sudo docker run --rm hello-world
+    if [ $? -ne 0 ]; then
+        echo "Docker installation failed. Please check the error messages above."
+        exit 1
+    fi
+    echo "Docker has been successfully installed."
```
Comment (ching-kuo) [inline install-linux.sh:55]: If we use `https://get.docker.com/` we can probably support other distro.

File: install-linux.sh [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+# Function to install Node.js
+install_nodejs() {
+    echo "Node.js not found. Installing Node.js version 20..."
+    curl -sL https://deb.nodesource.com/setup_20.x | sudo -E bash -
+    sudo apt-get install -y nodejs
```
Comment (ching-kuo) [inline install-linux.sh:62]: [nvm](https://github.com/nvm-sh/nvm) might work if we want to support distro other than Ubuntu in a single shell script, but I'm not very familiar with node.js

**Review by ching-kuo — COMMENTED** [review]
Some comment on possible solutions to support more distro, not tested though.

**Comment by justinh-rahb** [comment]
> Awesome! I can try this with my Ubuntu Laptop and maybe also check the functionality with my dual boot Debian.

I think it _should_ work on Debian (and other derivatives of it and Ubuntu), *but your mileage may vary* lol

File: install-linux.sh (hunk 1 above)
Comment (justinh-rahb) [reply install-linux.sh:55]: Ooh nice, didn't know this existed. I wanted to do the same with Ollama's installer script too, so may as well do both this way and not have to maintain more code than necessary.

File: install-linux.sh (hunk 2 above)
Comment (justinh-rahb) [reply install-linux.sh:62]: I'm not familiar with `nvm`, I'll look into it though. Also considered Bun, and might still try that.

File: install-linux.sh [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    echo "Installing Open WebUI with Docker..."
+    docker pull ghcr.io/open-webui/open-webui:latest
+    docker run -d -p 3000:8080 --add-host=host.docker.internal:host-gateway -v open-webui:/app/backend/data --name open-webui --restart always ghcr.io/open-webui/open-webui:latest
+    echo "Open WebUI has been installed and started with Docker. Access it at http://localhost:3000"
+}
+
+# Main script starts here
+echo "Starting the Open WebUI installation script..."
+
+if [ "$NO_DOCKER" -eq 1 ]; then
```
Comment (jannikstdl) [inline install-linux.sh:115]: I had this issue in debian:

```
Starting the Open WebUI installation script...
install-linux.sh: line 115: [: : integer expression expected
Checking for Docker...
Docker could not be found. Would you like to install Docker? (y/n)
y
```
Did you have that on Ubuntu?

For me this was a fix:
`if [ -n "$NO_DOCKER" ]`

This checks if the var ist non empty and if true evaluates the then statement.
I think i am getting this error bc if you don't set the -`-no-docker` flag, the var doesn't exist and can't be checked to be 1.

**Review by jannikstdl — COMMENTED** [review]
I tested it on Debian.

Docker installation failed:

```
Reading state information... Done
Package docker-ce is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

Package docker-ce-cli is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'docker-ce' has no installation candidate
E: Package 'docker-ce-cli' has no installation candidate
E: Unable to locate package containerd.io
E: Couldn't find any package by glob 'containerd.io'
E: Couldn't find any package by regex 'containerd.io'
sudo: docker: command not found
Docker installation failed. Please check the error messages above.
(base) jannik@Jannik-PC-debian:~/Downloads$ 
```
Maybe we could try this with https://get.docker.com as @ching-kuo mentioned.

Installation with --no-docker flag works just fine!
In future if we host this script on openwebui.com we should check if the user is in the cloned repo on his machine, and if not clone it first.
Maybe something like this:
```
if [ -f ".git/config" ] && grep -q "https://github.com/open-webui/open-webui.git" .git/config; then
  HAS_REPO=true
else
  HAS_REPO=false
fi

if [ "$HAS_REPO" = false ]; then
  read -p "Enter the path to clone the repository (default is current directory): " CLONE_PATH
  CLONE_PATH="${CLONE_PATH:-.}"
  CLONE_PATH=$(eval echo "$CLONE_PATH")
  git clone https://github.com/open-webui/open-webui.git "$CLONE_PATH/open-webui"
  cd "$CLONE_PATH/open-webui"
fi
```

**Comment by justinh-rahb** [comment]
Thanks for the testing @jannikstdl I'm going to work some more on this tonight and incorporate yours and @ching-kuo's excellent feedback 👏

File: install-linux.sh [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+# Function to check if Docker is installed
+check_docker() {
+    echo "Checking for Docker..."
+    if ! command -v docker &> /dev/null; then
+        echo "Docker could not be found. Would you like to install Docker? (y/n)"
+        read -r choice
+        if [ "$choice" = "y" ]; then
+            install_docker
+        else
```
[the remark this answers is not in the document]
Comment (justinh-rahb) [reply install-linux.sh:25]: Good idea, will incorporate this when I get time to return to working on this, sorry for the delay!

[the remark this answers is not in the document]
Comment (phidauex) [reply install-linux.sh:25]: Thanks! Also, as some additional positive feedback, the containerless install did actually complete successfully and start the Open-WebUI service. Now, I didn't intend to run it on my Proxmox/Debian host, but it was a good demonstration.

**Comment by justinh-rahb** [comment]
I'm getting back to work on this today, sorry it's been taking forever everyone :)

**Comment by jannikstdl** [comment]
All good, we all have stuff to do.

Do you want to change the target branch from open-webui:main to :dev or do you mean the dev branch on your fork?

**Comment by justinh-rahb** [comment]
### To-Do:

1. **Update Existing Repository**:
   - If the Open WebUI repository already exists, perform a `git pull` to update it rather than cloning afresh.
   - Reinstall dependencies before executing docker-less installations to ensure all components are up-to-date.

2. **Ollama Environment Configuration**:
   - Set up and configure necessary environment variables for Ollama, especially for host and CORS origins, during installation from the script.

3. **Flexible Host Configuration**:
   - Allow specification of whether Ollama and Open WebUI are hosted on the same machine or different hosts.
   - Introduce interactive prompts during installation to configure the host location for Ollama.
   - Set environment variables based on user input to handle different hosting scenarios.

**Comment by justinh-rahb** [comment]
Alright I'm starting to get into perfectionist mode now so I'd really love it if someone would tell me to stop now LOL 😂

**Comment by justinh-rahb** [comment]
Just realized the systemd file needs a bit more work... no handling of Python venv there.

**Comment by tjbck** [comment]
@justinh-rahb on fire today 🚀🚀

**Comment by tjbck** [comment]
Perhaps it's better to close this and open a new PR now that `pip install` PR has been merged?

**Comment by justinh-rahb** [comment]
> Perhaps it's better to close this and open a new PR now that `pip install` PR has been merged?

Agreed, I'll do a major refac on it.

**Pull request closed without merging**